### PR TITLE
Prevent msg_id reuse, add use-mock-crypto feature, improve tests

### PR DESCRIPTION
**Pull request by madadam**

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @maqi

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: tests/maid_manager.rs [hunk 1]
```diff
@@ -667,3 +667,73 @@ fn account_concurrent_insert_key_put_data() {
...
+
+    let mut successes = 0;
+    while let Ok(event) = client.try_recv() {
+        if let Event::Response { response: Response::PutIData { res: Ok(()), .. }, .. } = event {
+            successes += 1;
+        }
+    }
+
+    // Only one request does succeed.
+    assert_eq!(successes, 1);
```
Comment (maqi) [inline tests/maid_manager.rs]: could there be chance that receives no response from network? 
i.e. half passed data0 req to DM and half passed data1, hence no accumulation?

File: src/personas/maid_manager/mod.rs [hunk 2]
```diff
@@ -137,19 +137,27 @@ impl MaidManager {
...
         }
 
         // Forwarding the request to NAE Manager.
-        let fwd_src = dst;
-        let fwd_dst = Authority::NaeManager(*data.name());
-        trace!("MM forwarding PutIData request to {:?}", fwd_dst);
-        routing_node
-            .send_put_idata_request(fwd_src, fwd_dst, data, msg_id)?;
-        self.insert_into_request_cache(msg_id, src, dst, None);
+        if self.insert_into_request_cache(msg_id, src, dst, None) {
```
Comment (maqi) [inline src/personas/maid_manager/mod.rs]: function of insert_into_request_cache will replace the old one anyway.
so actually here we are not only invalidates the current request but also mismatch the response of old request from DM with the current request
I think here we shall carry out a `has` check then carry on the operations?

and same to the other places

Comment (madadam) [inline src/personas/maid_manager/mod.rs]: Fixed. `insert_into_request_cache` won't overwrite previous entries anymore.

File: src/personas/maid_manager/mod.rs [hunk 3]
```diff
@@ -1163,6 +1172,14 @@ enum PutMDataAction {
     Forward(MutableData),
 }
 
+struct RequestCacheInsert<'a>(VacantEntry<'a, MessageId, CachedRequest>, CachedRequest);
+
+impl<'a> RequestCacheInsert<'a> {
+    fn commit(self) {
```
Comment (ustulation) [inline src/personas/maid_manager/mod.rs]: ah nicely done :+1:

**Pull request merged**